### PR TITLE
Allow optional host parameter when mounting apps

### PR DESCRIPTION
**Pull request by robbl-as**
This pull-request introduces the possibility to mount an application using an optional host parameter. IMHO this closes a gap to the implementation of hanami/router, that allows the host parameter.

```ruby
Hanami.configure do
  mount(Api::Application, host: 'api.myapp.org', at: '/v1')
  mount(Web::Application, host: 'myapp.org', at: '/v1')
end
```

### CODE REVIEW

**Comment by robbl-as** [comment]
This pull-request also addresses the issue #656.

File: lib/hanami/configuration.rb [hunk 1]
```diff
@@ -42,7 +43,7 @@ def initialize(&blk)
     #     # ...
     #   end
     def mount(app, options)
-      mounted[app] = App.new(app, options.fetch(:at))
+      mounted[app] = App.new(app, options.fetch(:at), options.fetch(:host) { nil })
```
Comment (artofhuman) [inline lib/hanami/configuration.rb]: Can we use `options[:host]` ?
Should we add this option to documentation and add example?

Comment (robbl-as) [inline lib/hanami/configuration.rb]: @artofhuman Good point, I simplified the read access for the host option. Actually we got quite confused with the existing documentation http://hanamirb.org/guides/routing/overview/#mounting-on-a-subdomain. It implies, that there is already the possibility to mount an application with a given host. Prior to my pull-request the host options was ignored and not delegated to hanami/router.

**Comment by cllns** [comment]
Thanks for this!

What are `account` and `history` here? subdomains or full host names?  If they're just subdomains, I wonder if we should name the option `subdomain` instead? 

I could imagine a use-case of a single app on completely different host names. I'm not sure how we balance allowing people to mount apps on entirely different domains while letting people mount on a subdomain without having to hard-code the rest of the URL in the router. thoughts?

**Comment by robbl-as** [comment]
@cllns In my example account and history are complete hostnames. I am coming from a docker environment perspective, where hostnames like these are not unusual. Anyhow I updated my pull-request description. When referring to the complete hostname, mounting to subdomains is possibile, too, of cause.

**Comment by robbl-as** [comment]
Also I am not sure, why the travis build to this pull-request for the last commit failed. Is it possible to rerun the build?

**Comment by cllns** [comment]
great, thanks! Does providing a regex, instead of a string work too?

So you could mount to a subdomain without providing the rest of the domain?
```
Hanami.configure do
  mount(Api::Application, host: %r{^api.\w+}, at: '/v1')
  mount(Web::Application, at: '/v1')
end
```

**Comment by sandelius** [comment]
@cllns don't think so, what I can see it's just passed along to HttpRouter and there only a string is supported.

**Comment by robbl-as** [comment]
@cllns @sandelius There is no support for regex. I'd really like to see this pull request merged without an implementation that supports regex. In fact this pull request should be handled like a bugfix since the current documentation with in the hanami guides is not correct. The section http://hanamirb.org/guides/routing/overview/#mounting-on-a-subdomain describes mounting an application to a subdomain using the host-option. The host options is ignored. This pull-request fixes the issue.

**Comment by jodosha** [comment]
@robbl-as Hi and thanks for this PR.

---

Here's the steps I used to manually test it. I'm on Mac OS.

```
➜ pwd
/Users/luca/Code/hanami/hanami
➜ git checkout -b robbl-as-feature/allow-host-option-when-mounting-apps master
➜ git pull https://github.com/robbl-as/hanami.git feature/allow-host-option-when-mounting-apps
```

Then I generated a project that uses Hanami HEAD: https://github.com/hanami/hanami#how-to-use-hanami-head

I edited `/etc/hosts` and flushed the DNS cache:

```
127.0.0.1 bookshelf.local api.bookshelf.local # ...
```

```
➜ dscacheutil -flushcache
```

From the root of the test project:

```shell
➜ pwd
/Users/luca/Code/hanami/hanami/bookshelf
➜ bundle exec hanami generate app api
# ...
```

I edited `config/environment.rb` and started the server

```ruby
Hanami.configure do
  mount Api::Application, at: '/api', host: 'api.bookshelf.local'
  # ...
end
```

```shell
➜ bundle exec hanami server
# ...
```

At this point I was able to hit http://api.bookshelf.local:2300/

### Problem 1

At that URL I see the welcome page that tells me how to generate the first action for `Api` application:

```shell
➜ bundle exec hanami generate action web home#index --url=/
# ...
```

If I visit again that URL, **I still see the welcome page**. This is wrong.

If I inspect the routes, there is something wrong:

```shell
➜ bundle exec hanami routes
                Name Method     Path                           Action

                     GET, HEAD  /api                           Api::Controllers::Home::Index

                Name Method     Path                           Action

```

It looks like the URL that Hanami sees is http://api.bookshelf.local:2300/api, but we expect http://api.bookshelf.local:2300/. Indeed, by visiting http://api.bookshelf.local:2300/api, I can see the new generated page.

### Problem 2

If I generate a new action:

```shell
➜ bundle exec hanami generate action api books#index --url=/books
# ...
```

And I assign a name to that route:

```ruby
# apps/api/config/routes.rb
get '/books', to: 'books#index', as: :books
```

When I use the routes helper from a template it returns a wrong URL:

```erb
# apps/api/templates/home/index.html.erb
<%= routes.books_url %>
```

It returns http://localhost:2300/api/books, but we expect http://api.bookshelf.local:2300/books.

### Problem 3

If I add a stylesheet at `apps/api/assets/stylesheets/application.css` and refer it from `apps/api/templates/application.html.erb`:

```erb
<%= stylesheet "application" %>
```

It generates the wrong URL for it:

```html
<link href="/assets/api/application.css" type="text/css" rel="stylesheet">
```

---

Before to merge this we want to fix these problems above. Thank you.

**Comment by robbl-as** [comment]
@jodosha Thanks for diving in! Acctually I cannot recreate some of your problems, because it just works for me, even by following the same steps. Maybe I have to setup a repository with a test application we can work with? I had no issues with the routes or the assets.

The only issue I was able to reproduce: The routes url helper does not include the assigned host but still the localhost. I could look into this, though.

**Comment by TiteiKo** [comment]
I haven't tested it but I have a question...

ATM, assets path are generated based on the routing of the assets app. Meaning that if you have a `Web` app mounted at `/` and an `Admin` app mounted at `/admin`, `Web` assets paths will look like `/assets/application.css` and `Admin` assets paths will look like `/assets/admin/application.css`

With your change, if we have a `Web` app at `/` and an `Admin` app at `/` on a subdomain, both having an `application.css` asset, what will happen? What will be the assets paths?

**Comment by robbl-as** [comment]
@TiteiKo Looks like we have another issue. :) I tried what you suggested and both asset paths point to the same relative url `/assets/application.css`. Since we use the suggested changes from this pull-request for two API-only-applications mounted to `/api` but with different host-options we didn't run into any assets related issues. :smile: I failed to fix this issue by setting an explicit asset host for both applications. The assets paths are now legit `http://api.myhost.de/assets/application.css` but still not resolved to the correct file. Maybe this could be fixed somewhere [over here](https://github.com/hanami/hanami/blob/master/lib/hanami/assets/asset.rb#L35). Looks like assets are only resolved based on the path but not the host.

```
module Api
  class Application < Hanami::Application
    configure do
      host 'api.myhost.de'

      assets do
        host 'api.myhost.de'
      end
    end
  end
end
```

**Comment by jodosha** [comment]
Closing this as stale contribution. I'm marking it "for future consideration", because I'd like to see it in 2.0.

**Pull request closed without merging**